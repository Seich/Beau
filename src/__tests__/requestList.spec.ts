import Config from '../config'
import RequestList from '../requestList'
const requestPromiseNativeMock = require('request-promise-native')

describe('RequestList', () => {
    const endpoint = 'http://martianwabbit.com'

    let env = {
        environmental: true
    }

    const doc = {
        endpoint: endpoint,
        environment: env,
        'GET /post': { alias: 'get-posts' },
        'POST /user': {
            alias: 'user',
            payload: {
                name: 'David',
                lastname: 'Diaz'
            }
        }
    }

    let requests: RequestList
    beforeEach(() => {
        requestPromiseNativeMock.fail = false

        let config = new Config(doc)
        requests = new RequestList(config)
    })

    it('should allow an empty request list', () => {
        requests = new RequestList()
        expect(requests.list.length).toBe(0)
    })

    it('should load valid requests', () => {
        expect(requests.list.length).toBe(2)
    })

    it('should fetch dependencies', async () => {
        await expect(
            requests.fetchDependencies(['get-posts'])
        ).resolves.toMatchSnapshot()
    })

    it('should execute requests by alias.', async () => {
        await expect(requests.execByAlias('user')).resolves.toMatchSnapshot()
    })

    it('should fail if the request fails', async () => {
        requestPromiseNativeMock.fail = true
        await expect(requests.execByAlias('user')).rejects.toThrow()
    })

    it('should return a cached result if available', async () => {
        const obj = { test: true }
        requests.cache.add('test', obj)
        await expect(requests.execByAlias('test')).resolves.toBe(obj)
    })

    it('should fail if the alias is not found', async () => {
        await expect(requests.execByAlias('notAnAlias')).rejects.toThrow()
    })

    it(`should fail if a given request doesn't have an alias`, () => {
        let config = new Config({
            'GET /hello': {
                headers: {
                    hello: 1
                }
            }
        })

        expect(() => new RequestList(config, config)).toThrow()
    })
})
