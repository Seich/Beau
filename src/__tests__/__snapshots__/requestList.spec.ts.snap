// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestList should execute requests by alias. 1`] = `
Object {
  "body": "{\\"hello\\": \\"world\\"}",
  "request": Object {
    "body": Object {
      "lastname": "Diaz",
      "name": "David",
    },
    "endpoint": "http://martianwabbit.com/user",
    "headers": undefined,
  },
  "response": Object {
    "body": "{\\"hello\\": \\"world\\"}",
    "headers": Array [],
    "status": 200,
  },
}
`;

exports[`RequestList should fetch dependencies 1`] = `
RequestCache {
  "$cache": Object {
    "env": Object {
      "environmental": true,
    },
    "get-posts": Object {
      "body": "{\\"hello\\": \\"world\\"}",
      "request": Object {
        "body": undefined,
        "endpoint": "http://martianwabbit.com/post",
        "headers": undefined,
      },
      "response": Object {
        "body": "{\\"hello\\": \\"world\\"}",
        "headers": Array [],
        "status": 200,
      },
    },
  },
}
`;
